### DISCUSSION TITLE
Toolbars questions

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2117] [Q&A] [answered]
I added my own toolbar to our iTwin app, using the example frontend-sample>>toobar-button-provider-sample.  Here is a sample code snippet from my provideToolbarButtonItems()

       const toolsToolbar: CommonToolbarItem [] = [ 
        ToolbarHelper.createToolbarItemFromItemDef(1000, CoreTools.measureToolGroup),
        ToolbarItemUtilities.createActionButton("clearMeasure", 2000, "icon-measure-clear", "Clear Measurements", () => {})];  return toolsToolbar;

 I would like to more interactivity to the toolbar, similar to behavior of measure and clear measure buttons in iModel Design Review

![measure-toolbar](https://user-images.githubusercontent.com/83007711/130496504-6df909ab-7b47-4226-bfb8-0e4b3efa6b75.JPG)

The clear measurement button shows up only after user performs a measurement, otherwise it is hidden.  Other interactivity I would like to add is to show that a button is active (similar to select button is highlighted indicating that it is active) and if possible to show that button is 'pressed' to act like a toggle.

Is there an example code I can look at which implements such behavior (perhaps in one of iTwin modules).

#2)  I want to provide a toolbar button to clear measurements, as is in Design Review.  I created an ActionButton in my toolbar (see code snippet above) but I was unable to find a command or (core?) tool that I can call to clear them.  Please guide.

#3) I want to set opacity of all toolbars, similar to Design Review (as in pic above).  Without the background, toolbar buttons tend to get lost as in the pic below.  Visibility of toolbar buttons is much better in Design Review because the toolbar background is always shown.  Please guide.

![LostToolbar](https://user-images.githubusercontent.com/83007711/130498074-23c95dfa-0074-48a3-bc64-e065f2458b04.JPG)

**Comment by aruniverse** [comment]
Re conditionally showing a toolbar item: 
- [ToolbarHelper.createToolbarItemFromItemDef](https://www.itwinjs.org/reference/ui-framework/toolbar/toolbarhelper/) takes an arg to override [ToolbarItem](https://www.itwinjs.org/reference/ui-abstract/toolbar/toolbaritem/) properties; what you want is probably `isDisabled` and/or `isHidden`. 
- A good example can be found in the measure-tools package used by design review, see [here](https://github1s.com/imodeljs/viewer-components-react/blob/master/packages/measure-tools/src/ui-2.0/MeasureToolsUiProvider.tsx#L83-L84)

**Comment by bsteinbk** [comment, marked as answer]
#2 The "Core" measure tools do not come with a clear measurement command. If you press the right mouse button while the measure tools are active the existing measurement decorations will be erased. If you want to simulate the button in Design Review you could try adding the following to your UiItemsProvider.
```
    if (toolbarUsage === ToolbarUsage.ContentManipulation && toolbarOrientation === ToolbarOrientation.Horizontal) {
      const isClearMeasureHiddenCondition = new ConditionalBooleanValue((): boolean =>
         !IModelApp.toolAdmin.currentTool?.toolId.startsWith("Measure."), [SyncUiEventId.ToolActivated]);
      const clearMeasureActionSpec = ToolbarItemUtilities.createActionButton("clear-measure-tool", -100, 
        "icon-paintbrush", "Clear Measure Decorations",
        (): void => {
          IModelApp.toolAdmin.currentTool?.onReinitialize();
        }, { isHidden: isClearMeasureHiddenCondition });
      return [clearMeasureActionSpec]
    }
```
#3 Add the following function to make the toolbar opacity level static instead of changing as the pointer gets closer to the toolbar.
```
function onInit () {
  UiShowHideManager.useProximityOpacity = false;
}
```
And then call that function during inIModelAppInit processing for the Viewer component
```
 onIModelAppInit={onInit}
```

**Reply by ovaisa** [reply]
Thank you Bill, I was able to resolve #2 and #3, the provided example was very helpful in gaining understanding of toolbar interaction.   

I tried to play around with isPressed in similar way you showed how to use isHidden but no much luck.  isPressed is a boolean (as compared to ConditionalBoolean) so I tried setting it to true/false but it didn't seem to change icon look.  If there is and example of how isPressed is set, I would be very helpful  (as it is readOnly so I can set it on the object)

@Arun, I haven't gotten around to #1 but thank you for pointer to code, I plan to look at it in next day or two...  Much appreciated!

**Comment by bsteinbk** [comment]
At some point in the development history we decided to move away from supporting toggle buttons in the toolbars and that is why `isPressed` is not working. It was decided that having both active and pressed buttons would be confusing. The current scheme is to change the icon based on state. Below is an example on how you would set that up using a ConditionalStringValue to specify the icon.

```
      const conditionalIcon = new ConditionalStringValue(() => {
        return myToggleManager.isActive ? "icon-camera-animation" : "icon-camera-animation-disabled";
      }, ["my-toggle-state-changed"]),

      const myToggleActionSpec = ToolbarItemUtilities.createActionButton("my-toggle-tool", -50, 
        conditionalIcon, "Toggle My Setting",
        (): void => {
          myToggleManager.isActive = !myToggleManager.isActive;
		  // tell UI to refresh items registered to monitor my event
		  SyncUiEventDispatcher.dispatchSyncUiEvent("my-toggle-state-changed");
        });
```

**Reply by ovaisa** [reply]
Makes sense, I was also uncertain if I should show pressed state or active to avoid confusion.  I now see that many other properties in ItemProps are marked as Depricated, recommend that isPressed should also be marked Depricated. 

Another suggestion on API documentation.  I have found in few cases that same class/interface is present in 2 modules.  I ended up importing the wrong module (or VSCode picked the wrong one) causing confusion.  It think it is rare but has tripped me more than couple of times.  Also it would be helpful if the API documentation indicated which module should be imported for a class/interface (it is not always obvious and difficult to figure out)

![Duplicates](https://user-images.githubusercontent.com/83007711/130651023-1830a139-6537-49ba-8d7d-42e7584e65db.JPG)


Thanks for providing this code, seems it should achieve my objective.

**Comment by ovaisa** [comment]
Hi Bill,

Thanks for providing guidance on popup panel in toolbar items, this is exactly what I needed.  Much appreciated!!!!

One question I had was that I couldn't find a way to specify callback for the toolbar icon itself.  I need to capture state of a variable when the icon is pressed so that the slider can manipulate it.  Is there a callback I can specify for the icon or callback when panel open (and closes) - couldn't find in CustomToolbarItem.  

One way to achieve this would be to develop a React class or function and use it as panelConentNode (rather than Slider),  I can then use useEffect() or one of React lifecycle callbacks but I wanted to check perhaps if there is a way to avoid developing this class.

Thanks again, much appreciate your guidance.  I have few more questions but I will post them later as a new thread...

**Reply by bsteinbk** [reply]
We currently don't provide any callbacks for open/close. The only thing we provide is a method for you to call to close the popup panel.  So your best bet is to making your own component similar to what I show below.

```
function MySliderPanel() {
  const [sliderValues, setSliderValues] = React.useState([50]);
  React.useEffect(() => {
    // eslint-disable-next-line no-console
    console.log("mounting my panel");
    return () => {
      // eslint-disable-next-line no-console
      console.log("unmounting my panel");
    };
  }, []);

  const { closePanel } = useToolbarPopupContext();
  const handleApply = React.useCallback(() => {
    closePanel();
  }, [closePanel]);

  const handleChange = React.useCallback((values) => {
    setSliderValues(values);
  }, []);

  return (
    <div style={{ display: "flex", flexDirection: "column", gap: "4px", alignItems: "center", width: "300px", height: "68px", padding: "6px", boxSizing: "border-box" }}>
      <Slider style={{ width: "100%" }} min={0} max={100} values={sliderValues} step={1} onChange={handleChange} />
      <Button onClick={handleApply}>Apply</Button>
    </div>
  );
}

```
